File: daemon/internal/protocol/varint.go
package protocol

// maxVarintBytes bounds LEB128 decoding so a crafted prefix can never
// allocate an unbounded integer. 5 bytes covers uint32, the widest field.
const maxVarintBytes = 5

// AppendVarint encodes v as unsigned LEB128. Callers must keep v < 2^32.
func AppendVarint(dst []byte, v uint64) []byte {
	for {
		b := byte(v & 0x7f)
		v >>= 7
		if v != 0 {
			b |= 0x80
		}
		dst = append(dst, b)
		if v == 0 {
			return dst
		}
	}
}

// ReadVarint decodes one unsigned LEB128 from the front of b. It reports the
// value and the number of bytes consumed.
func ReadVarint(b []byte) (uint64, int, error) {
	var v uint64
	for i := 0; i < len(b); i++ {
		if i == maxVarintBytes {
			return 0, 0, ErrVarintTooLong
		}
		c := b[i]
		v |= uint64(c&0x7f) << (7 * i)
		if c&0x80 == 0 {
			return v, i + 1, nil
		}
	}
	return 0, 0, ErrVarintTruncated
}

File: app/android/app/src/main/java/com/remotly/app/transport/Varint.kt
package com.remotly.app.transport

// Unsigned LEB128, capped at 5 bytes (the widest protocol field is a uint32
// channel id). Mirrors the daemon's protocol.AppendVarint / ReadVarint byte for
// byte, so the app and daemon agree on framing.
object Varint {
    const val MAX_BYTES = 5

    // v must be < 2^32.
    fun encode(v: Long): ByteArray {
        require(v in 0..0xFFFFFFFFL) { "varint out of range: $v" }
        var value = v
        val out = ByteArray(MAX_BYTES)
        var i = 0
        while (true) {
            var b = (value and 0x7f).toInt()
            value = value ushr 7
            if (value != 0L) b = b or 0x80
            out[i++] = b.toByte()
            if (value == 0L) break
        }
        return out.copyOf(i)
    }

    // Returns the decoded value and the number of bytes consumed. Throws on a
    // truncated or over-long prefix.
    fun decode(b: ByteArray, off: Int): Pair<Long, Int> {
        var v = 0L
        var i = off
        while (true) {
            val pos = i - off
            if (pos == MAX_BYTES) throw IllegalArgumentException("varint too long")
            if (i >= b.size) throw IllegalArgumentException("varint truncated")
            val c = b[i].toInt() and 0xff
            v = v or ((c and 0x7f).toLong() shl (7 * pos))
            i++
            if (c and 0x80 == 0) return v to (i - off)
        }
    }
}
